File: js/viginere.js
const cycle = (arr) => [...arr, ...arr, ...arr]

const vigenere = (alphabet, key, message) => {
    return message.split("").map((char, index) => {
        const charIndex = cycle(alphabet).indexOf(char)
        return charIndex != -1 ? cycle(alphabet)[charIndex + key[index]]: char
    })
}

let rersult = vigenere(["a", "b", "c", "d", "e", "f"], [1, 2, 1, 2, 1], "abcdf")
console.log(rersult)
/*
vigenere :: [Char] -> [Int] -> String -> String
vigenere alphabet key message = undefined

Использовать 
cycle - слепливающийся массив [1,2,3] = [1,2,3,1,2,3,1,2,3....]
elemIndex = indexOf
(!!) - получение элемента массива по индексу
zipWith
*/

File: haskell/viginere.hs
import Data.List

vigenere :: [Char] -> [Int] -> String -> String
vigenere alphabet keys message = do
    zipWith shiftChar message (cycle keys)

    where
        shiftChar char key = if getIndexOfCharInAlphabet char /= -1 
            then (alphabet !! (mod ((getIndexOfCharInAlphabet char) + key) (length alphabet)))
            else char
        getIndexOfCharInAlphabet char = case elemIndex char alphabet of
            Just n  -> n
            Nothing -> -1

decodeVigenere alphabet keys message = vigenere alphabet (map negate keys) message    

main :: IO ()
main = do 
    let result = vigenere ['a', 'b', 'c'] [-4, 10] "abc"
    let decodedMessage = decodeVigenere ['a', 'b', 'c', 'd', 'e', 'f'] [-1, 2] result
    print ("Encoded to: " ++ result)
    print ("Decoded to: " ++ decodedMessage)

